File: srcs/api-cpp/RawProperty.cpp
/*
 * This file is part of Nokia HEIF library
 *
 * Copyright (c) 2015-2018 Nokia Corporation and/or its subsidiary(-ies). All rights reserved.
 *
 * Contact: heif@nokia.com
 *
 * This software, including documentation, is protected by copyright controlled by Nokia Corporation and/ or its
 * subsidiaries. All rights are reserved. Copying, including reproducing, storing, adapting or translating, any or all
 * of this material requires the prior written consent of Nokia.
 */

#include "RawProperty.h"
#include <heifreader.h>
#include <heifwriter.h>
#include <cstring>

using namespace HEIFPP;

RawProperty::RawProperty(Heif* aHeif)
    : ItemProperty(aHeif, HEIF::ItemPropertyType::RAW, HEIF::FourCC((std::uint32_t)0), false)
{
}
RawProperty::RawProperty(Heif* aHeif, const HEIF::FourCC& aType, bool aIsTransform)
    : ItemProperty(aHeif, HEIF::ItemPropertyType::RAW, aType,aIsTransform)
{
}

RawProperty::~RawProperty()
{
}
void RawProperty::getData(const std::uint8_t*& aData, std::uint64_t& aLength) const
{
    aData   = mRaw.data.elements + 8;
    aLength = mRaw.data.size - 8;
}
void RawProperty::setData(const std::uint8_t* aData, std::uint64_t aLength)
{
    const int offset = 8;//offset to payload (size of box type and size)
    mRaw.data    = HEIF::Array<uint8_t>((aLength + offset));
    mRaw.data[0] = ((aLength + offset) >> 24) & 0xFF;
    mRaw.data[1] = ((aLength + offset) >> 16) & 0xFF;
    mRaw.data[2] = ((aLength + offset) >> 8) & 0xFF;
    mRaw.data[3] = ((aLength + offset)) &0xFF;    
    mRaw.data[4] = static_cast<std::uint8_t>(rawType().value[0]);
    mRaw.data[5] = static_cast<std::uint8_t>(rawType().value[1]);
    mRaw.data[6] = static_cast<std::uint8_t>(rawType().value[2]);
    mRaw.data[7] = static_cast<std::uint8_t>(rawType().value[3]);
    std::memcpy(mRaw.data.elements + offset, aData, aLength);
}
HEIFPP::Result RawProperty::setRawType(const HEIF::FourCC& aType, bool aIsTransform)
{
    HEIFPP::Result res=ItemProperty::setRawType(aType);
    if (res == HEIFPP::Result::OK)
    {
        setIsTransformative(aIsTransform);
        // TODO: if mIsTransform changes AFTER the property is associated with an item,
        // we should move it to correct place in the props list. (declarative properties first, then transformative).
    }
    return res;
}
HEIF::ErrorCode RawProperty::load(HEIF::Reader* aReader, const HEIF::PropertyId& aId)
{
    HEIF::ErrorCode error;
    error = ItemProperty::load(aReader, aId);
    if (HEIF::ErrorCode::OK == error)
    {
        aReader->getProperty(aId, mRaw);
        if (mRaw.data.size < 8)  // needs atleast 8 bytes
        {
            return HEIF::ErrorCode::MEDIA_PARSING_ERROR;
        }
        std::uint32_t size;
        size = static_cast<uint32_t>(mRaw.data[0]) << 24;
        size |= static_cast<uint32_t>(mRaw.data[1]) << 16;
        size |= static_cast<uint32_t>(mRaw.data[2]) << 8;
        size |= static_cast<uint32_t>(mRaw.data[3]);
        HEIF::FourCC fc;
        fc.value[0] = static_cast<char>(mRaw.data[4]);
        fc.value[1] = static_cast<char>(mRaw.data[5]);
        fc.value[2] = static_cast<char>(mRaw.data[6]);
        fc.value[3] = static_cast<char>(mRaw.data[7]);
        fc.value[4] = 0;
        setRawType(fc,false);
        if (size != mRaw.data.size)
        {
            return HEIF::ErrorCode::MEDIA_PARSING_ERROR;
        }        
        if (rawType() != mRaw.type)
        {
            return HEIF::ErrorCode::MEDIA_PARSING_ERROR;
        }
    }
    return error;
};
HEIF::ErrorCode RawProperty::save(HEIF::Writer* aWriter)
{
    HEIF::ErrorCode error = HEIF::ErrorCode::OK;
    HEIF::PropertyId newId;
    error = aWriter->addProperty(mRaw, isTransformative(), newId);
    if (HEIF::ErrorCode::OK == error)
        setId(newId);
    return error;
}

File: srcs/api-java/java/com/nokia/heif/RawProperty.java
/*
 * This file is part of Nokia HEIF library
 *
 * Copyright (c) 2015-2018 Nokia Corporation and/or its subsidiary(-ies). All rights reserved.
 *
 * Contact: heif@nokia.com
 *
 * This software, including documentation, is protected by copyright controlled by Nokia Corporation and/ or its subsidiaries. All rights are reserved.
 * Copying, including reproducing, storing, adapting or translating, any or all of this material requires the prior written consent of Nokia.
 *
 *
 */

package com.nokia.heif;

import java.nio.ByteBuffer;

/**
 * Rawproperty
 */
public final class RawProperty extends ItemProperty
{

    /**
     * Creates a RawProperty to the given HEIF instance
     * @param heif The parent HEIF instance for the new object
     * @throws Exception Thrown if the parent HEIF instance is invalid
     */
    public RawProperty(HEIF heif, FourCC type)
            throws Exception
    {
        super(heif);
        mNativeHandle = createContextNative(heif);
        try
        {
            setRawType(type);
        }
        catch (Exception ex)
        {
            destroy();
            throw ex;
        }
    }

    /**
     * Protected constructor, used to create an object from the native side
     * @param heif The parent HEIF instance for the new object
     * @param nativeHandle A handle to the corresponding C++ object
     */
    protected RawProperty(HEIF heif, long nativeHandle)
    {
        super(heif, nativeHandle);
    }

    /**
     * Returns the type of the property
     * @return Type as a FourCC
     * @throws Exception
     */
    public FourCC getRawType()
            throws Exception
    {
        checkState();
        return new FourCC(getRawTypeNative());
    }

    public byte[] getDataAsByteArray()
            throws Exception
    {
        ByteBuffer buffer = getData();
        byte[] byteArray = new byte[buffer.remaining()];
        buffer.get(byteArray);
        return byteArray;
    }

    public ByteBuffer getData()
            throws Exception
    {
        checkState();
        ByteBuffer output = getDataNative();
        return output.asReadOnlyBuffer();
    }

    public void setData(byte[] data)
            throws Exception
    {
        checkState();
        checkParameter(data);
        setDataNative(data);
    }

    private void setRawType(FourCC type)
            throws Exception
    {
        checkState();
        checkParameter(type);
        setRawTypeNative(type.toString());
    }

    private native long createContextNative(HEIF heif);

    private native String getRawTypeNative();

    private native void setRawTypeNative(String type);

    private native ByteBuffer getDataNative();

    private native void setDataNative(byte[] data);



}
